test/metabase/test/redefs.clj

```diff
@@ -5,22 +5,28 @@
    [mb.hawk.parallel]
    [metabase.plugins.classloader :as classloader]
    [methodical.core :as methodical]
+   [toucan2.connection :as t2.connection]
    [toucan2.tools.with-temp :as t2.with-temp]))
 
+(def ^{:dynamic true
+       :private true}
+  *in-with-temp* false)
+
 (methodical/defmethod t2.with-temp/do-with-temp* :around :default
-  "Initialize the DB before doing the other with-temp stuff. Make sure metabase.test.util is loaded. Make sure we are
-  not inside a parallel test."
+  "Initialize the DB before doing the other with-temp stuff. Make sure metabase.test.util is loaded.
+  Make sure the the "
   [model attributes f]
   (classloader/require 'metabase.test.initialize)
   ((resolve 'metabase.test.initialize/initialize-if-needed!) :db)
   ;; so with-temp-defaults are loaded
   (classloader/require 'metabase.test.util)
-  ;; disallow with-temp inside parallel tests.
-  ;;
-  ;; TODO -- there's not really a reason that we can't use with-temp in parallel tests -- it depends on the test -- so
-  ;; once we're a little more comfortable with the current parallel stuff we should remove this restriction.
-  (mb.hawk.parallel/assert-test-is-not-parallel "with-temp")
-  (next-method model attributes f))
+
+  ;; run `f` in a transaction if it's the top-level with-temp
+  (if-not *in-with-temp*
+    (binding [*in-with-temp* true]
+      (t2.connection/with-transaction [_]
+        (next-method model attributes f)))
+    (next-method model attributes f)))
```

Comment:
I think we still need some sort of escape hatch to do `with-temp` outside of a transaction for tests that make real HTTP requests on a separate thread. I think maybe we want to add a dynamic variable like

```clj
(def ^:dynamic *with-temp-use-transactions* true)
```

and then add a thread-unsafe macro called `with-temp!` like this

```clj
(defmacro with-temp! [& args]
  `(binding [*with-temp-use-transactions* false]
     (t2.with-temp/with-temp ~@args)))
```

Reply:
yea I created a `with-temp!` macro for this and another helper `mt/with-ensure-with-temp-no-transaction!` to use at the top-level